xapian-core/weight/tfidfweight.cc

```diff
@@ -136,7 +142,8 @@ TfIdfWeight::get_maxpart() const
 {
     Xapian::termcount wdf_max = get_wdf_upper_bound();
     Xapian::termcount len_min = get_doclength_lower_bound();
-    double wdfn = get_wdfn(wdf_max, len_min, normalizations[0]);
+    Xapian::termcount len_max = get_doclength_upper_bound();
+    double wdfn = get_wdfn(wdf_max, len_min, len_max, normalizations[0]);
     return get_wtn(wdfn * idfn, normalizations[2]) * wqf_factor;
```

Comment:
`get_maxpart()` needs to return an upper bound over all documents in the collection, but `uniqterms` varies per document.  Note how we use the **bounds** on `doclen`, we don't pass in `doclen` itself because there is no appropriate single `doclen` value - the same applies to `uniqterms`.

We don't track bounds on `uniqterms` currently, but I don't think we can do better than evaluating at `doclen == uniqterms` unless `doclen_min >= uniqterms_max`.

(Actually, it's not quite true that `uniqterms <= doclen` since a term in a document can have wdf=0 but such terms would still count in the unique count - this is an esoteric corner case, and probably best dealt with by setting `doclen = max(doclen, uniqterms)` (or something equivalent) when calculating the `L` norm value).